src/libs/ajax.js

```diff
@@ -58,14 +58,60 @@ const withErrorRejection = wrappedFetch => async (...args) => {
   }
 }
 
+const withUrlPrefix = prefix => wrappedFetch => async (path, ...args) => {
+  return wrappedFetch(prefix + path, ...args)
+}
+
+const checkRequesterPaysError = async response => {
+  if (response.status === 400) {
+    const data = await response.json()
+    const requesterPaysError = _.includes('requester pays', _.get(['error', 'message'], data))
+    return Object.assign(new Response(new Blob([JSON.stringify(data)]), response), { requesterPaysError })
+  } else {
+    return Object.assign(response, { requesterPaysError: false })
+  }
+}
+
+const mergeQueryParams = (params, urlString) => {
+  const url = new URL(urlString)
+  url.search = qs.stringify({ ...qs.parse(url.search, { ignoreQueryPrefix: true, plainObjects: true }), ...params })
+  return url.href
+}
+
+/*
+ * Detects errors due to requester pays buckets, and adds the current workspace's billing
+ * project if the user has access, retrying the request once if necessary.
+ */
+const withRequesterPays = wrappedFetch => async (url, ...args) => {
+  const bucket = /\/b\/([^/]+)\//.exec(url)[1]
+  const workspace = workspaceStore.get()
+  const userProject = workspace && Utils.canWrite(workspace.accessLevel) && workspace.workspace.namespace
+  const tryRequest = async () => {
+    const knownRequesterPays = _.includes(bucket, requesterPaysBuckets.get())
+    try {
+      return await wrappedFetch(mergeQueryParams({ userProject: (knownRequesterPays && userProject) || undefined }, url), ...args)
+    } catch (error) {
+      const newResponse = await checkRequesterPaysError(error)
+      if (newResponse.requesterPaysError && !knownRequesterPays) {
+        requesterPaysBuckets.update(_.union([bucket]))
+        if (userProject) {
+          return tryRequest()
```

Comment:
Are we guaranteed that the `update` above will have completed before this call to `tryRequest` runs?

Reply:
Yes, atom updates are synchronous.